### PR TITLE
[feat]: LPL rule integrated into skeleton network

### PR DESCRIPTION
**Pull request by and-rewsmith**
Integrating the LPL rule into the skeleton network.  This runs successfully, but still need to assess selectivity metric.

Changes:

1. Fixed some bugs with datagen. Namely the cluster generation needs to start randomly from any potential cluster rather than starting automatically from 0th. Also started applying some normalization.
2. Added the temporal filters to the network. We are using one epsilon filter and _two_ separate alpha filters. 

```
        self.alpha_filter_first_term = TemporalFilter(
            tau_rise=TAU_RISE_ALPHA, tau_fall=TAU_FALL_ALPHA)
        self.epsilon_filter_second_term = TemporalFilter(
            tau_rise=TAU_RISE_EPSILON, tau_fall=TAU_FALL_EPSILON)
        self.alpha_filter_second_term = TemporalFilter(
            tau_rise=TAU_RISE_ALPHA, tau_fall=TAU_FALL_ALPHA)
```
3. Implemented the LPL rule. When implementing the LPL rule, there is a discrepancy with the first hidden layer, which is that it relies on spike, mem, and variance of previous layer spiking. This can be solved by Poisson encoding the input data, as well as applying our spike and variance filters to the incoming input data for our first hidden layer. For example, this is in the network init, which is responsible for tracking these intermediary values for the learning rule.
```
        self.data_spike_moving_average = SpikeMovingAverage(
            batch_size=layer_settings.batch_size, data_size=self.layer_settings.data_size)
        self.variance_moving_average = VarianceMovingAverage()
```

### CODE REVIEW

File: model/src/model.py [hunk 1]
```diff
@@ -32,37 +61,49 @@ def __init__(self,
...
         self.mem = self.forward_lif.init_leaky()
 
-        self.optimizer = torch.optim.Adam(
-            self.forward_weights.parameters(), lr=layer_settings.learning_rate)
+        self.alpha_filter_first_term = TemporalFilter(
+            tau_rise=TAU_RISE_ALPHA, tau_fall=TAU_FALL_ALPHA)
+        self.epsilon_filter_second_term = TemporalFilter(
+            tau_rise=TAU_RISE_EPSILON, tau_fall=TAU_FALL_EPSILON)
+        self.alpha_filter_second_term = TemporalFilter(
+            tau_rise=TAU_RISE_ALPHA, tau_fall=TAU_FALL_ALPHA)
```
Comment (and-rewsmith) [inline model/src/model.py:131]: Note: we are using two alpha filters.

File: model/src/util.py [hunk 2]
```diff
@@ -1,22 +1,34 @@
...
 from typing import Any, Optional
 import snntorch as snn
 import torch
 
+MAX_RETAINED_SPIKES = 2
 
-class MovingAverageLIF(snn.LIF):
-    def __init__(self, *args: Any, tau_mean: float, tau_var: float, **kwargs: Any) -> None:
+# Zenke's paper uses a tau_mean of 600s
+TAU_MEAN = 600
```
Comment (and-rewsmith) [inline model/src/util.py]: Calling this out again, but scaling the units in a certain way will cause different interactions with the learning rule. If we hope to pick the same values as Zenke and hope it mostly works, this may need to be tuned.

File: model/src/model.py [hunk 3]
```diff
@@ -79,20 +120,93 @@ def forward(self, data: Optional[torch.Tensor] = None) -> tuple[torch.Tensor, to
...
+
+         3. The third term contains S_j. We form a matrix via some unsqueezes to
+            form a third term matrix of size (batch_size, j, i). This is mainly
+            done by expanding the tensor to duplicate values for i dimension.
+
+        The final dw_ij/dt is formed by a Hadamard product of the first term and
+        the second term, and then adding the third term. This is then summed
+        across the batch dimension and divided by the batch size to form the
+        final dw_ij/dt matrix. We then apply this to the weights.
+        """
```
Comment (and-rewsmith) [inline model/src/model.py:262]: FYA to see this note

File: model/src/util.py [hunk 4]
```diff
@@ -99,8 +113,14 @@ def __init__(self, tau_mean: float = 600) -> None:
         """
         self.mean: Optional[torch.Tensor] = None
         self.tau_mean = tau_mean
+        self.spike_rec: Deque[torch.Tensor] = deque(maxlen=MAX_RETAINED_SPIKES)
```
Comment (and-rewsmith) [inline model/src/util.py:119]: Since the `SpikeMovingAverage` abstraction was already processing spikes, I have this deque in here to store the last two spikes. 

This is good because it doesn't need to be tracked outside, however causes a bit of weirdness because the mem still needs to be tracked. That has been left in `Layer`. So now they are in two places, but I think it is fine.

File: model/src/model.py [hunk 5]
```diff
@@ -80,19 +121,95 @@ def forward(self, data: Optional[torch.Tensor] = None) -> tuple[torch.Tensor, to
...
+            # third term
+            third_term = self.layer_settings.learning_rate * DELTA * \
+                self.forward_lif.spike_moving_average.spike_rec[-1]
+            prev_layer_size = self.layer_settings.data_size if self.prev_layer is None \
+                else self.prev_layer.layer_settings.size
+            third_term = third_term.unsqueeze(
+                2).expand(-1, -1, prev_layer_size)
+
+            # update weights
+            dw_dt = first_term * second_term_alpha + third_term
```
Comment (and-rewsmith) [inline model/src/model.py]: FYA: `first_term * second_term_alpha` is hadamard product

File: model/src/model.py [hunk 6]
```diff
@@ -80,19 +121,95 @@ def forward(self, data: Optional[torch.Tensor] = None) -> tuple[torch.Tensor, to
             current = self.forward_weights(data)
 
         spk, mem = self.forward_lif(current, self.mem)
-        self.spk_rec.append(spk)
         self.mem_rec.append(mem)
 
         return spk, mem
 
     def train_forward(self, data: Optional[torch.Tensor] = None) -> None:
```
Comment (and-rewsmith) [inline model/src/model.py:239]: I am mulling over if some white-box tests would be helpful here. Would need a refactor I think.

File: model/src/model.py [hunk 7]
```diff
@@ -32,37 +61,49 @@ def __init__(self,
...
         super().__init__()
 
+        self.layer_settings = layer_settings
+
         # weights from prev layer to this layer
         self.forward_weights = nn.Linear(
             layer_settings.prev_size, layer_settings.size)
-        self.forward_lif = snn.Leaky(
-            beta=layer_settings.beta, learn_beta=layer_settings.learn_beta)
+        self.forward_lif = MovingAverageLIF(batch_size=layer_settings.batch_size, layer_size=layer_settings.size,
```
Comment (and-rewsmith) [inline model/src/model.py:118]: In a separate PR we are going to need to add inhibitory neurons as well.

File: model/src/model.py [hunk 8]
```diff
@@ -1,13 +1,42 @@
...
 import snntorch as snn
+from snntorch import spikegen
 
 from datasets.src.zenke_2a.constants import TEST_DATA_PATH, TRAIN_DATA_PATH
+from datasets.src.zenke_2a.datagen import generate_sequential_dataset
 from datasets.src.zenke_2a.dataset import SequentialDataset
+from model.src.util import MovingAverageLIF, SpikeMovingAverage, TemporalFilter, VarianceMovingAverage
+
+# Zenke's paper uses a theta_rest of -50mV
+THETA_REST = -50
```
Comment (and-rewsmith) [inline model/src/model.py]: This likely doesn't make any sense as a value given snnTorch having 0 as lowest mem?

File: model/src/model.py [hunk 9]
```diff
@@ -32,37 +61,49 @@ def __init__(self,
...
-        self.forward_lif = snn.Leaky(
-            beta=layer_settings.beta, learn_beta=layer_settings.learn_beta)
+        self.forward_lif = MovingAverageLIF(batch_size=layer_settings.batch_size, layer_size=layer_settings.size,
+                                            beta=layer_settings.beta, learn_beta=layer_settings.learn_beta)
 
-        # TODO: cap size
-        self.spk_rec: List[torch.Tensor] = []
-        self.mem_rec: List[torch.Tensor] = []
+        self.mem_rec: Deque[torch.Tensor] = Deque(maxlen=MAX_RETAINED_MEMS)
+        for _ in range(2):
```
Comment (atran39) [inline model/src/model.py]: use MAX_RETAINED_MEMS here instead of 2?

**Comment by and-rewsmith** [comment]
I am playing with this now in order to get a measure of selectivity. There is some weird behavior with weight updates turning NaN after some number of iterations. Trying to fix.

File: model/src/model.py [hunk 10]
```diff
@@ -1,28 +1,58 @@
...
 import snntorch as snn
+from snntorch import spikegen
 
 from datasets.src.zenke_2a.constants import TEST_DATA_PATH, TRAIN_DATA_PATH
+from datasets.src.zenke_2a.datagen import generate_sequential_dataset
 from datasets.src.zenke_2a.dataset import SequentialDataset
+from model.src.util import MovingAverageLIF, SpikeMovingAverage, TemporalFilter, VarianceMovingAverage
+
+# Zenke's paper uses a theta_rest of -50mV
+THETA_REST = 0
```
Comment (mstormbull) [inline model/src/model.py:18]: noting our discussion from earlier today: decision point between 1) custom forward, 2) using Auryn, 3) rescale and shift SNNtorch to match paper hyperparameters.

File: model/src/model.py [hunk 11]
```diff
@@ -80,19 +121,130 @@ def forward(self, data: Optional[torch.Tensor] = None) -> tuple[torch.Tensor, to
...
-
-        self.optimizer.step()
+        """
+        The LPL learning rule is implemented here. It is defined as dw_ji/dt,
+        for which we optimize the computation with matrices.
+
+        This learning rule is broken up into three terms:
+
+         1. The first term contains S_j(t) * f'(U_i(t)). We form a matrix via
+            some unsqueezes to form a first term matrix of size (batch_size, j,
```
Comment (mstormbull) [inline model/src/model.py]: Confirming that this should be an outer product which takes {Nx1] [1xN] = [NxN].

File: model/src/model.py [hunk 12]
```diff
@@ -80,19 +121,130 @@ def forward(self, data: Optional[torch.Tensor] = None) -> tuple[torch.Tensor, to
...
+            i). This is mainly done by matrix multiplication.
+
+         2. The second term contains S_i. We form a matrix via some unsqueezes
+            to form a second term matrix of size (batch_size, i, j). This is
```
Comment (mstormbull) [inline model/src/model.py:251]: confirming that this should be a row duplication procedure.

File: model/src/model.py [hunk 13]
```diff
@@ -80,19 +121,130 @@ def forward(self, data: Optional[torch.Tensor] = None) -> tuple[torch.Tensor, to
...
+            mainly done by expanding the tensor to duplicate values for j
+            dimension.
+
+         3. The third term contains S_j. We form a matrix via some unsqueezes to
+            form a third term matrix of size (batch_size, j, i). This is mainly
```
Comment (mstormbull) [inline model/src/model.py]: column duplication this time.

File: model/src/model.py [hunk 14]
```diff
@@ -80,19 +121,130 @@ def forward(self, data: Optional[torch.Tensor] = None) -> tuple[torch.Tensor, to
...
+            done by expanding the tensor to duplicate values for i dimension.
+
+        The final dw_ij/dt is formed by a Hadamard product of the first term and
```
Comment (mstormbull) [inline model/src/model.py:258]: 100% correct

File: model/src/model.py [hunk 15]
```diff
@@ -80,19 +121,130 @@ def forward(self, data: Optional[torch.Tensor] = None) -> tuple[torch.Tensor, to
...
+            if data is not None:
+                data_mean = self.data_spike_moving_average.apply(data)
+                _variance = self.data_variance_moving_average.apply(
+                    spike=data, spike_moving_average=data_mean)
+
+            spk, _mem = self.forward(data)
+
+            # first term
+            prev_layer_mem = torch.ones(
+                self.layer_settings.batch_size, self.layer_settings.data_size) * DATA_MEM_ASSUMPTION \
```
Comment (mstormbull) [inline model/src/model.py]: I can double check this once we flip pre-post again, per discussion from today.

File: model/src/model.py [hunk 16]
```diff
@@ -129,37 +281,52 @@ def __init__(self, settings: Settings) -> None:
...
         for epoch in range(self.settings.epochs):
             for i, batch in enumerate(train_loader):
+                # permute to (num_steps, batch_size, data_size)
                 batch = batch.permute(1, 0, 2)
-                batch = batch[:, :, :1]
 
-                print(
-                    f"Epoch {epoch} - Batch {i} - Sample data: {batch.shape}")
+                # poisson encode
+                spike_trains = spikegen.rate(batch, time_var_input=True)
```
Comment (mstormbull) [inline model/src/model.py]: might not be needed for the initial two cluster data set.

**Review by mstormbull — CHANGES_REQUESTED** [review]
Aside from the Pre-post flip and the associated double check, this LGTM.

File: model/src/model.py (hunk 11 above)
Comment (and-rewsmith) [reply model/src/model.py]: yep that is correct

File: model/src/model.py (hunk 12 above)
Comment (and-rewsmith) [reply model/src/model.py:251]: yep that is correct

File: model/src/model.py (hunk 13 above)
Comment (and-rewsmith) [reply model/src/model.py]: correct

File: model/src/model.py (hunk 16 above)
Comment (and-rewsmith) [reply model/src/model.py]: I agree, however it is a bit smelly that in that case S_j would be a continuous value. I was thinking it wasn't that big of a deal, however offline I thought of how S_j can actually be a continuous value above the spike upper limit of 1. This seems bad enough to warrant always Poisson encoding?

File: model/src/model.py (hunk 11 above)
Comment (and-rewsmith) [reply model/src/model.py]: Noted this in the comment

File: model/src/model.py (hunk 13 above)
Comment (and-rewsmith) [reply model/src/model.py]: Noted this in the comment

File: model/src/model.py (hunk 7 above)
Comment (mstormbull) [reply model/src/model.py:118]: We will need inhibitory neurons to 'balance' the network and not have it blow up. Stefano has thought a lot about this so might be fun to chat with him about the role between dimensionality and balance.

File: model/src/model.py [hunk 17]
```diff
@@ -80,19 +148,203 @@ def forward(self, data: Optional[torch.Tensor] = None) -> tuple[torch.Tensor, to
...
+        with torch.no_grad():
+            if data is not None:
+                spk, mem = self.forward(data)
+            else:
+                spk, mem = self.forward()
+
+            # first term
+            f_prime_u_i = ZENKE_BETA * \
+                (1 + ZENKE_BETA * abs(self.mem - THETA_REST)) ** (-2)
+            f_prime_u_i = f_prime_u_i.unsqueeze(2)
```
Comment (mstormbull) [inline model/src/model.py:272]: ✔️ expanded along the correct axis -- broadcasting all post synaptic to presynaptic.

File: model/src/model.py [hunk 18]
```diff
@@ -80,19 +148,203 @@ def forward(self, data: Optional[torch.Tensor] = None) -> tuple[torch.Tensor, to
...
+            prev_layer_most_recent_spike: torch.Tensor = self.prev_layer.forward_lif.spike_moving_average.spike_rec[
+                0] if self.prev_layer is not None else data  # type: ignore [union-attr, assignment]
+            prev_layer_most_recent_spike = prev_layer_most_recent_spike.unsqueeze(
+                1)
```
Comment (mstormbull) [inline model/src/model.py:276]: ✔️

File: model/src/model.py [hunk 19]
```diff
@@ -80,19 +148,203 @@ def forward(self, data: Optional[torch.Tensor] = None) -> tuple[torch.Tensor, to
...
+            first_term_no_filter = prev_layer_most_recent_spike * f_prime_u_i
```
Comment (mstormbull) [inline model/src/model.py:277]: this shape should be batch x neuron x neuron

File: model/src/model.py [hunk 20]
```diff
@@ -80,19 +148,203 @@ def forward(self, data: Optional[torch.Tensor] = None) -> tuple[torch.Tensor, to
...
+            current_layer_variance_moving_average = self.forward_lif.variance_moving_average.tracked_value()
+
+            second_term_prediction_error = current_layer_most_recent_spike - \
+                current_layer_delta_t_spikes_ago
+            second_term_deviation_scale = LAMBDA_HEBBIAN / \
+                (current_layer_variance_moving_average + XI)
+            second_term_deviation = current_layer_most_recent_spike - \
+                current_layer_spike_moving_average
+            second_term_no_filter = -1 * \
+                (second_term_prediction_error) + second_term_deviation_scale * \
```
Comment (mstormbull) [inline model/src/model.py:298]: ✔️

File: model/src/model.py [hunk 21]
```diff
@@ -80,19 +148,203 @@ def forward(self, data: Optional[torch.Tensor] = None) -> tuple[torch.Tensor, to
...
+                second_term_no_filter)
+            second_term_alpha = second_term_alpha.unsqueeze(
+                2).expand(-1, -1, self.layer_settings.data_size)
+
+            # third term
+            prev_layer_most_recent_spike: torch.Tensor = self.prev_layer.forward_lif.spike_moving_average.spike_rec[
+                0] if self.prev_layer is not None else data  # type: ignore [union-attr, assignment]
+            third_term = DELTA * prev_layer_most_recent_spike
+            third_term = third_term.unsqueeze(
+                1).expand(-1, self.layer_settings.size, -1)
```
Comment (mstormbull) [inline model/src/model.py:310]: ✔️

File: model/src/model.py [hunk 22]
```diff
@@ -80,19 +148,203 @@ def forward(self, data: Optional[torch.Tensor] = None) -> tuple[torch.Tensor, to
...
+
+            # update weights
+            dw_dt = self.layer_settings.learning_rate * first_term_alpha * \
+                second_term_alpha + self.layer_settings.learning_rate * third_term
```
Comment (mstormbull) [inline model/src/model.py]: minor: might be cleaner to write as: 
`dw_dt = self.layer_settings.learning_rate * (first_term_alpha * second_term_alpha + third_term)`

did I miss something here?

File: model/src/model.py [hunk 23]
```diff
@@ -80,19 +148,203 @@ def forward(self, data: Optional[torch.Tensor] = None) -> tuple[torch.Tensor, to
...
+            dw_dt = dw_dt.sum(0) / dw_dt.shape[0]
```
Comment (mstormbull) [inline model/src/model.py:315]: ✔️

**Review by mstormbull — APPROVED** [review]
LGTM. Great job with those LR expansions1

File: model/src/model.py (hunk 22 above)
Comment (and-rewsmith) [reply model/src/model.py]: Will fix then merge
